### PR TITLE
More logging for IO with explicit duration, class, and thread names

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

File: core/src/main/scala/cromwell/core/io/DefaultIoCommand.scala [hunk 1]
```diff
@@ -9,22 +9,43 @@ object DefaultIoCommand {
                                   override val destination: Path,
                                   override val overwrite: Boolean) extends IoCopyCommand(
     source, destination, overwrite
-  )
-  case class DefaultIoContentAsStringCommand(override val file: Path, override val options: IoReadOptions) extends IoContentAsStringCommand(file, options)
-  case class DefaultIoSizeCommand(override val file: Path) extends IoSizeCommand(file)
+  ) {
+    customDebug(s"DefaultIoCopyCommand.init source '$source' destination '$destination' overwrite '$overwrite'")
+  }
```
Comment (mcovarr) [inline core/src/main/scala/cromwell/core/io/DefaultIoCommand.scala:14]: TOL are we expecting to see any of these `Default` guys or just the `Gcs` variants?

File: wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala [hunk 2]
```diff
@@ -325,7 +331,20 @@ object WdlStandardLibraryFunctions {
 
         // Inner function: Get the file size, allowing for unpacking of optionals
         def optionalSafeFileSize(value: WomValue): Try[Long] = value match {
-          case f if f.isInstanceOf[WomSingleFile] || WomSingleFileType.isCoerceableFrom(f.womType) => Try(Await.result(ioFunctionSet.size(f.valueString), Duration.Inf))
+          case f if f.isInstanceOf[WomSingleFile] || WomSingleFileType.isCoerceableFrom(f.womType) => {
+            val uuid = UUID.randomUUID().toString
+            val creation = OffsetDateTime.now
+            logger.info(f"($uuid) (0.000 seconds) optionalSafeFileSize $value start ($getClass) (thread '${Thread.currentThread().getName}')")
```
Comment (mcovarr) [inline wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:337]: is `value` going to give us something as useful as `f.valueString`?

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala [hunk 3]
```diff
@@ -518,8 +524,20 @@ object EngineFunctionEvaluators {
...
-          f.coerceToType[WomSingleFile] flatMap { file => Try(Await.result(ioFunctionSet.size(file.valueString), Duration.Inf)).toErrorOr }
+        case f if f.isInstanceOf[WomSingleFile] || WomSingleFileType.isCoerceableFrom(f.womType) => {
+          val uuid = UUID.randomUUID().toString
+          val creation = OffsetDateTime.now
+          logger.info(f"($uuid) (0.000 seconds) optionalSafeFileSize $value start ($getClass) (thread '${Thread.currentThread().getName}')")
+          val result = f.coerceToType[WomSingleFile] flatMap { file => Try(Await.result(ioFunctionSet.size(file.valueString), Duration.Inf)).toErrorOr }
+          val now = OffsetDateTime.now
+          val millis = java.time.Duration.between(creation, now).toMillis
+          val seconds = millis / 1000D
+          logger.info(f"($uuid) ($seconds%,.3f seconds) optionalSafeFileSize $value done ($getClass) (thread '${Thread.currentThread().getName}')")
```
Comment (mcovarr) [inline wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:535]: same `value` / `f.valueString` question here

File: core/src/main/scala/cromwell/core/io/DefaultIoCommand.scala (hunk 1 above)
Comment (kshakir) [reply core/src/main/scala/cromwell/core/io/DefaultIoCommand.scala:14]: I ran it locally and with my config I see defaults.

Comment (mcovarr) [reply core/src/main/scala/cromwell/core/io/DefaultIoCommand.scala:14]: interesting and not what I expected, I'll have a look at that...

Comment (kshakir) [reply core/src/main/scala/cromwell/core/io/DefaultIoCommand.scala:14]: Screen Shot 2020-10-28 at 1 35 23 PM

Here's one example trace. And my WDL:

```WDL
version 1.0

task t {
    input { File in_file }
    Float in_file_size = size(in_file)
    command <<< echo '~{in_file}' '~{in_file_size}' >>>
    runtime { docker: "ubuntu" }
    output {
        String out = read_string(stdout())
        Float out_size = size(stdout())
    }
}

workflow w {
  call t { input: in_file = "gs://wb-mock-drs-dev/public/some_data.txt" }
  output {
      String out = t.out
      Float out_size = t.out_size
  }
}
```

File: wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala (hunk 2 above)
Comment (kshakir) [reply wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala]: switched

File: wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala (hunk 3 above)
Comment (kshakir) [reply wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala]: switched

File: filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchIoCommand.scala [hunk 4]
```diff
@@ -121,16 +124,19 @@ sealed trait GcsBatchGetCommand[T] extends SingleFileGcsBatchIoCommand[T, Storag
 }
 
 case class GcsBatchSizeCommand(override val file: GcsPath, setUserProject: Boolean = false) extends IoSizeCommand(file) with GcsBatchGetCommand[Long] {
+  customDebug(s"GcsBatchSizeCommand.init '$file'")
```
Comment (mcovarr) [inline filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchIoCommand.scala:127]: were these `setUserProject`s intentionally omitted here and below?

File: engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchCommandContext.scala [hunk 5]
```diff
@@ -73,14 +73,21 @@ final case class GcsBatchCommandContext[T, U](request: GcsBatchIoCommand[T, U],
...
+    request.customDebug(s"GcsBatchCommandContext.queue '$batchRequest'")
+    request.operation.queue(batchRequest, callback)
+  }
 
   /**
     * On success callback. Transform the request response to a stream-ready response that can complete the promise
     */
-  private def onSuccessCallback(response: U, httpHeaders: HttpHeaders) = handleSuccessOrNextRequest(request.onSuccess(response, httpHeaders))
+  private def onSuccessCallback(response: U, httpHeaders: HttpHeaders) = {
+    request.customDebug(s"GcsBatchCommandContext.onSuccessCallback '$response'")
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchCommandContext.scala:85]: TOL good call not logging the headers 😉

File: core/src/main/scala/cromwell/core/io/DefaultIoCommand.scala (hunk 1 above)
Comment (kshakir) [reply core/src/main/scala/cromwell/core/io/DefaultIoCommand.scala:14]: My conf is a mess (I need to clean out a lot of old / temp config), but here it is JIC:

<details>
<summary>application.jes.conf (may be easier to view/copy in github's edit commend mode)</summary>
<pre>
include required("application")

workflow-options.workflow-failure-mode: "ContinueWhilePossible"

backend {
  default = "Papi"
  backendsAllowed = [ "Papi" ]

  providers {
    Local.config.filesystems.gcs.auth = "application-default"
    Papi {
      name = "Papi"
      actor-factory = "cromwell.backend.google.pipelines.v2alpha1.PipelinesApiLifecycleActorFactory"
      config {
        #concurrent-job-limit = 2

        // Google project
        project = "broad-dsde-cromwell-dev"

        // Base bucket for workflow executions
        root = "gs://kshakir-dsde-cromwell-dev/cromwell-executions"

        // Polling for completion backs-off gradually for slower-running jobs.
        // This is the maximum polling interval (in seconds):
        maximum-polling-interval = 600

        genomics {
          // A reference to an auth defined in the ``google`` stanza at the top.  This auth is used to create
          // Pipelines and manipulate auth JSONs.
          auth = "application-default"
          // Endpoint for APIs, no reason to change this unless directed by Google.
          endpoint-url = "https://genomics.googleapis.com/"
        }

        filesystems = {
          gcs.auth = "application-default"
//          drs.auth = "drs-user"
//          http.enabled: true
        }
      }
    }
  }
}

google {
  applicationName = "cromwell-kshakir"
  auths = [
    {
      name = "application-default"
      scheme = "application_default"
    }
    #{
    #  name = "user-service-account"
    #  scheme = "user_service_account"
    #}
    #{
    #  name = "service-account"
    #  scheme = "service_account"
    #  json-file = "/Users/kshakir/src/cromwell/cromwell-service-account.json"
    #}
//    {
//      name = "drs-user"
//      scheme = "user_account"
//      secrets-file = ${HOME}"/.config/gcloud/legacy_credentials/baron.von.testerton@gmail.com/adc.json"
//      secrets-file = ${HOME}"/.config/gcloud/legacy_credentials/kshakir@broadinstitute.org/adc.json"
//    }
  ]
}

call-caching.enabled = true
#call-caching.invalidate-bad-cache-results = false

engine {
  filesystems {
    gcs.auth = "application-default"
//    drs.auth = "drs-user"
    http.enabled: true
  }
}

system {
  new-workflow-poll-rate = 1
}

filesystems.drs.global.config.martha.url = "https://us-central1-broad-dsde-dev.cloudfunctions.net/martha_v3"

system.delete-workflow-files = true

drs.localization.docker-image = "broadinstitute/cromwell-drs-localizer:develop"

# Turning this on while debuging PROD-444
system.file-hash-cache = true
</pre>
</details>

File: core/src/main/scala/cromwell/core/io/IoCommand.scala [hunk 6]
```diff
@@ -21,21 +25,46 @@ object IoCommand {
...
+  private val uuid = UUID.randomUUID().toString
+  private val creation = OffsetDateTime.now
+
+  def customDebug(message: String): Unit = {
+    val now = OffsetDateTime.now
+    val millis = java.time.Duration.between(creation, now).toMillis
+    val seconds = millis / 1000D
+    IoCommand.logger.info(f"($uuid) ($seconds%,.3f seconds) $message ($getClass) (thread '${Thread.currentThread().getName}')")
+    if (seconds > 5 * 60) {
+      IoCommand.logger.warn(f"($uuid) ($seconds%,.3f seconds) FYI! `$message` IO IS OVER 5 MINUTES ($getClass) (thread '${Thread.currentThread().getName}')")
```
Comment (cjllanwarne) [inline core/src/main/scala/cromwell/core/io/IoCommand.scala:43]: FWIW I often find `getClass.getSimpleName` more useful/readable

Comment (cjllanwarne) [inline core/src/main/scala/cromwell/core/io/IoCommand.scala:43]: This message may be a little clunky but 🤷

Comment (kshakir) [reply core/src/main/scala/cromwell/core/io/IoCommand.scala:43]: True: but in this case I'm overlogging and want the whole package

File: engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchCommandContext.scala (hunk 5 above)
Comment (cjllanwarne) [reply engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchCommandContext.scala:85]: Do we potentially want a `toPrettyElidedString` on the response?

File: core/src/main/scala/cromwell/core/io/IoCommand.scala (hunk 6 above)
Comment (kshakir) [reply core/src/main/scala/cromwell/core/io/IoCommand.scala:43]: Happy to have better english. I just want something I can grep for quickly in the logs, then backtrack from there using the `uuid`.

File: filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchIoCommand.scala (hunk 4 above)
Comment (kshakir) [reply filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/batch/GcsBatchIoCommand.scala]: added

File: wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala [hunk 7]
```diff
@@ -301,6 +305,8 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WomValue] {
 }
 
 object WdlStandardLibraryFunctions {
+  private val logger = LoggerFactory.getLogger(WdlStandardLibraryFunctions.getClass)
```
Comment (cjllanwarne) [inline wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:308]: Is this better that mixing in `with StrictLogging`?

File: wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala [hunk 8]
```diff
@@ -325,7 +331,20 @@ object WdlStandardLibraryFunctions {
 
         // Inner function: Get the file size, allowing for unpacking of optionals
         def optionalSafeFileSize(value: WomValue): Try[Long] = value match {
-          case f if f.isInstanceOf[WomSingleFile] || WomSingleFileType.isCoerceableFrom(f.womType) => Try(Await.result(ioFunctionSet.size(f.valueString), Duration.Inf))
+          case f if f.isInstanceOf[WomSingleFile] || WomSingleFileType.isCoerceableFrom(f.womType) => {
+            val uuid = UUID.randomUUID().toString
```
Comment (cjllanwarne) [inline wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:335]: TOL: For this and the other UUIDs in this PR, is there a way we could format them differently so they're harder to confuse with other UUIDs in the logs?

**Review by cjllanwarne — APPROVED** [review]
Lots of TOLs but no show-stoppers

File: wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala (hunk 8 above)
Comment (kshakir) [reply wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:335]: I put `()` around them. Very different than the others surrounded by `[]`. 😉 

If you have some other prefix/suffix suggestion I'm happy to swap it out.

File: engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchCommandContext.scala (hunk 5 above)
Comment (kshakir) [reply engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchCommandContext.scala:85]: I want the whole response. For now. I may regret that later.

File: wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala (hunk 7 above)
Comment (kshakir) [reply wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:308]: I got some cryptic compiler error when trying to use scala-logging pointing to this previous `import` line:

```
/Users/kshakir/src/cromwell/wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:12:36
encountered unrecoverable cycle resolving import.
Note: this is often due in part to a class depending on a definition nested within its companion.
If applicable, you may wish to try moving some members into another object.
import wdl.draft2.model.expression.WdlStandardLibraryFunctions.{crossProduct => stdLibCrossProduct, _}
```

Don't feel like figuring it for now.

**Review by mcovarr — APPROVED** [review]

**Review by grsterin — APPROVED** [review]

**Comment by kshakir** [comment]
Merging with caveats:
- sbt tests are passing
- aws centaur is using an old queue
- other centaurs are failing due to expected and excessive logging

None of this is expected to negatively affect production

**Pull request merged**